### DISCUSSION TITLE
Is there a way to run one model/condition per GPU when using multiple GPUs?

### DISCUSSION THREAD

**Discussion by li-li-github** [optuna/optuna#4255] [Q&A]
https://github.com/optuna/optuna-examples/blob/main/pytorch/pytorch_lightning_ddp.py
The above example uses multiple GPUs for data parallelization where multiple GPUs are working on the same model at a time.

I wonder if there's a way to run one model per GPU so that I can examine multiple models/conditions with multiple GPUs at the same time. I'm asking this question because I feel running one model per GPU at a time can be more efficient than data parallelization. For instance, in the above example code, FashionMNIST is a small dataset that doesn't take up much GPU memory, so data parallelization doesn't make much sense. Instead, running one model per GPU or even multiple models per GPU should better utilize GPU memories. 

Thank you.

**Comment by gen740** [comment]
I'm not sure what you want to do exactly, but if you want to utilize only one GPU. Just use another method other than this complicated feature.

Could you tell us more information about models, (and setups) for more detailed advice?

**Reply by li-li-github** [reply]
Sorry for the unclearness. 

For example, I have 4 GPUs and  want to search for the best learning rate between 1e-2 to 1e-5.

I was thinking if I could train a model with `lr=1e-2` on GPU0, another model with `lr=1e-3` on GPU1, another model with `lr=1e-4` on GPU2 and another model with `lr=1e-5` on GPU3 at the same time.
In such a way, I could examine multiple parameters at the same time on multiple GPUs.

In the given example code, it seems to start with training a model with `lr=1e-2` on all 4 GPUs, and then train a model with `lr=1e-3` on all 4 GPUs, and then train a model with `lr=1e-4` on all 4 GPUs and so on and so forth.
In summary, the code seems always training one model at a time using all GPUs. This may not be the most efficient way to use all the GPUs especially when the dataset and model is not so big like the ones given in the example code. It only occupies less than 2GB of GPU memory with 1 GPU, and using 4 GPUs wouldn't give much benefits (it was using ~0.7GB on each GPU).

In my practical use case, this is also important because I don't want to increase my batch size when I use 4 GPUs. In other words, I want to find the optimal set of parameters for a small batch size (that can fit in 1 GPU) by using 4 GPUs to speedup the searching process.
I wonder what would be the best way to achieve my aim.

Let me know if there's anything I couldn't make clear enough!

**Reply by gen740** [reply]
Sorry for the late reply.

I guess the easiest way to accomplish this is using optuna's [multiprocess feature](https://optuna.readthedocs.io/en/stable/tutorial/10_key_features/004_distributed.html). And assign  a GPU to each process.
